### ISSUE TITLE
Directive/transform does not work as expected by using common parameters in the OpenAPI doc

### ISSUE THREAD

**Issue by tibortakacs** [Azure/autorest#3516]
**Summary**
The OpenAPI spec allows to use common parameters in the description (https://swagger.io/docs/specification/describing-parameters/ - Common Parameters section). However, if the `directive` and `transform` tags are used in the `readme.md` AutoRest config in order to transform the operationIds, the generation successfully finishes but without the expected transformation.

Example:
**readme_gwd.md**
```
# PetStore
> see https://aka.ms/autorest
This is the AutoRest configuration file for the APIs.
---
#### Basic Information
These are the global settings for APIs.
``` yaml
# list all the input OpenAPI files (may be YAML, JSON, or Literate- OpenAPI markdown)
input-file:
  - gatewayd_simple.yaml
# this allows you to programatically tweak the swagger file before it is modeled.
directive:
  from: openapi-document # do it globally
  where: $.paths.*.*
  # set each operationId to 'MyAPI<Tag>'
  transform: $.operationId = `${$.tags[0]}_${$.operationId}`
```

**gatewayd_simple.yaml**
```
openapi: "3.0.0"

info:
  version: 4.1.0
  title: Azure Sphere Device REST API
  description: REST API exposed by Azure Sphere devices via the Servicing Port

servers:
  - url: https://192.168.35.2

paths:
  /abi_versions:
    get:
      operationId: "getAbiVersions"
      tags:
        - "images"
      summary: "Get ABI versions"
      responses:
        200:
          description: "ABI version set"
          content:
            application/json:
              schema:
                type: string
  /app/status/{componentId}:
    parameters:
      - in: path
        name: componentId
        required: true
        description: "Component ID of application"
        schema:
          type: string
    get:
      tags:
        - "app"
      operationId: "getAppStatus"
      summary: "Get app status"
      responses:
        200:
          description: "App status"
          content:
            application/json:
              schema:
                type: string
```
**Autorest script**
`autorest --v3 --python --use=@autorest/python@5.1.0-preview.1 .\readme_gwd.md --debug`

The `/app/status/{componentId}` endpoint has a common parameter. The interesting point is that none of the transformation works not just there where the common parameter is used.

**Comment by tibortakacs** [comment]
@daviwil I finished the investigation, please find my analysis here:
Under the path `/app/status/{componentId}`, there are two direct nodes in the tree: `parameters` and `get`. The `where: $.paths.*.*` fit for both. The execution of the directive in the `core\lib\pipeline\object-manipulator.ts` `manipulateObject(...)` function:
* First, the `where` clause is executed:
https://github.com/Azure/autorest/blob/9855de29b5c335b0dead2c0a8dccf8b3eb604b64/core/lib/pipeline/object-manipulator.ts#L38
* The `hits` list contains two elements, one for the `get` (first item) and one for the `parameters` node (second item), 
* Then, the actual transformations are executed:
https://github.com/Azure/autorest/blob/9855de29b5c335b0dead2c0a8dccf8b3eb604b64/core/lib/pipeline/object-manipulator.ts#L49
* First for the `get`. It works fine, and the tree is modified as expected.
* Then, for the the `parameters`. However, this node does not contain the `tags` and `operationId` fields which causes failure in the `transformer(...)` lambda execution.
* This failure causes an uncaught exception in this function which is caught on level above:
https://github.com/Azure/autorest/blob/9855de29b5c335b0dead2c0a8dccf8b3eb604b64/core/lib/pipeline/manipulation.ts#L87
* This part suppresses this issue, so no error or warning printed to the console, and it continues executing the pipeline with the original AST.
* So, the `operationId` is not overwritten.

I see two possible solutions here:
1. We let to fail the individual `transformer(...)` labmdas without interrupting the full process. Please find a draft PR here: https://github.com/Azure/autorest/pull/3525
2. The user has to write more complex `where` clause. (I am not sure, however, whether it is possible or not.)

**Comment by daviwil** [comment]
Thanks for the thorough investigation!  Let me look at this a little more closely and I'll get back to you.

**Comment by daviwil** [comment]
After trying a bunch of different things, it doesn't seem JSONPath allows you to narrow your search based on the names of properties in an object, so I don't think there's a way you could constrain the query further.  Since your scenario here (updating `operationIds`) is valid, I think we should probably allow a failing transformation so that the whole pipeline doesn't get stopped.  I think we should write out a warning to the console when this happens so that the user has a clue that something went wrong.

To write out a warning in this case, you'll need to add a call like this in the `catch` clause:

```javascript
config.Message({
 Channel: Channel.Warning,
 Details: err, // from `catch`
 Text: `Directive with 'where' clause '${whereJsonQuery}' failed to execute: ${err.message}`
});
```

To get a `config` instance you'll need to pass it into `manipulateObject` here: https://github.com/azure/autorest/blob/master/core/lib/pipeline/manipulation.ts#L38

**Comment by daviwil** [comment]
Over Teams @fearthecowboy shared a directive clause that would work in your case:

```
$.paths.*[?(@.tags)]*
```

That will filter out all paths properties that don't contain a `tags` property.  Perhaps you could try that?  Regardless of that, we'll probably still take the PR to swallow transform errors where they are applied so that it doesn't disrupt further transforms.

**Comment by tibortakacs** [comment]
@daviwil Thank you for your response (and sorry for the delay in my response).
Adding a warning message is a great idea, going to add to the PR soon.
I have tried the recommended query, unfortunately it does not work for me. The jsonpath module throws an exception during the parsing:
https://github.com/dchester/jsonpath/blob/778bc1ea297309e5925910698bb8ead9369d2e55/lib/index.js#L118
@fearthecowboy FYI

**Comment by daviwil** [comment]
Fixed by #3525.
